src/libraries/ReservoirLibrary.sol

```diff
@@ -0,0 +1,224 @@
+        require(aReserveIn > 0 && aReserveOut > 0, "RL: INSUFFICIENT_LIQUIDITY");
+        uint lNumerator = aReserveIn * aAmountOut * FEE_ACCURACY;
+        uint lDenominator = (aReserveOut - aAmountOut) * (FEE_ACCURACY - aSwapFee);
+        rAmountIn = (lNumerator / lDenominator) + 1;
+    }
+
+    function getAmountOutStable(
+        uint256 aAmountIn,
+        uint256 aReserveIn,
+        uint256 aReserveOut,
+        uint256 aSwapFee,
+        ExtraData memory aData
+    ) internal pure returns (uint256 rAmountOut) {
+        require(aAmountIn > 0, "RL: INSUFFICIENT_INPUT_AMOUNT");
+        require(aReserveIn > 0 && aReserveOut > 0, "RL: INSUFFICIENT_LIQUIDITY");
+
+        rAmountOut = StableMath._getAmountOut(
+            aAmountIn,
+            aReserveIn,
+            aReserveOut,
+            aData.token0PrecisionMultiplier,
+            aData.token1PrecisionMultiplier,
+            true,
+            aSwapFee,
+            2 * aData.amplificationCoefficient
+        );
+    }
+
+    function getAmountInStable(
+        uint256 aAmountOut,
+        uint256 aReserveIn,
+        uint256 aReserveOut,
+        uint256 aSwapFee,
+        ExtraData memory aData
+    ) internal pure returns (uint256 rAmountIn) {
+        require(aAmountOut > 0, "RL: INSUFFICIENT_OUTPUT_AMOUNT");
+        require(aReserveIn > 0 && aReserveOut > 0, "RL: INSUFFICIENT_LIQUIDITY");
+
+        rAmountIn = StableMath._getAmountIn(
+            aAmountOut,
+            aReserveIn,
+            aReserveOut,
+            aData.token0PrecisionMultiplier,
+            aData.token1PrecisionMultiplier,
+            false,
+            aSwapFee,
+            2 * aData.amplificationCoefficient
+        );
+    }
+
+    // performs chained getAmountOut calculations on any number of pairs
+    function getAmountsOut(
+        address aFactory,
+        uint256 aAmountIn,
+        address[] memory aPath,
+        uint256[] memory aCurveIds
+    ) internal view returns (uint256[] memory rAmounts) {
+        require(aPath.length >= 2, "RL: INVALID_PATH");
+        require(aCurveIds.length == aPath.length - 1, "RL: CURVE_IDS_INVALID_LENGTH");
+        rAmounts = new uint[](aPath.length);
+        rAmounts[0] = aAmountIn;
+        for (uint i = 0; i < aPath.length - 1; ) {
+            (uint lReserveIn, uint lReserveOut) = getReserves(aFactory, aPath[i], aPath[i + 1], aCurveIds[i]);
+            uint lSwapFee = getSwapFee(aFactory, aPath[i], aPath[i + 1], aCurveIds[i]);
+            if (aCurveIds[i] == 0) {
+                rAmounts[i + 1] = getAmountOutConstantProduct(rAmounts[i], lReserveIn, lReserveOut, lSwapFee);
+            }
+            else if (aCurveIds[i] == 1) {
+                ExtraData memory lData = ExtraData(
+                    uint64(getPrecisionMultiplier(aPath[i])),
+                    uint64(getPrecisionMultiplier(aPath[i + 1])),
+                    uint64(getAmplificationCoefficient(pairFor(aFactory, aPath[i], aPath[i + 1], 1)))
+                );
```

Comment:
getPrecisionMultiplier makes external calls to token right, so is quite expensive? We are also doing twice, so 2 CALL per token in path, let's add a flag to optimize this before code freeze?